### DISCUSSION TITLE
Questions on latency and resource usage?

### DISCUSSION THREAD

**Discussion by LeeNX** [Mystfit/ESP32-BLE-CompositeHID#1] [General]
Thanks @Mystfit , been poking at your repo and see a few pretty interesting things.  
  
I have a few questions around latency with bigger BLE packets (don't shoot me for the wrong terms). Have you done any testing on this front? Or are there tweaks or things to help improve latency?

**Comment by Mystfit** [comment]
Haven't done a lot of latency testing yet except for some perceptual latency tests as I'm building a split controller right now using two ESP32 boards (a satellite controller and a hub controller) and have been testing what the latency is from the satellite controller through the hub controller and then to the computer. I think that there is a very small amount of perceptual latency from the satellite controller since there are two jumps to the PC but I haven't tested the latency of a single controller. I can look into doing some sort of roundtrip latency tests using a light sensor.

I've been using a polling loop with a delay of about 8ms (little shorter than 120Hz update rate). Device reports are usually quite small, several bytes in size usually. The keyboard sends about 3 bytes per report and an emulated XBox controller sends around 16 bytes per report (not including any NimBLE bluetooth overheads).

Regarding latency, the default values I have set in BleConnectionStatus.cpp at the moment are  6 * 1.25ms minimum, connection interval 7ms * 1.25ms maximum connection interval, 0 latency (number of allowed skipped packets), 600 * 10ms timeout. I could make that configurable if you would like to configure your own connection parameters.

**Reply by LeeNX** [reply]
Thanks, that is alot of info!

I have been meaning to setup a testing process, where an app on something like a Raspberry Pi can trigger a button on the ESP32, which would trigger a button press over BLE and measure that, but RealLife(tm) keeps putting this off.

The reason I ask, is more about trying to understand the limitation of BLE. I mean, why does not every gamepad come with max button/sliders/axis and things connected. Yes, cost and pin limitation of base microcontroller. 

I just remembering in the older gaming console, adding more gamepad and hub, did seem to slow down some of the gamepads.

**Comment by LeeNX** [comment]
Come across the link before - https://gp2040-ce.info/#performance
